File: src/main/python/AoC2023_14.py
#! /usr/bin/env python3
#
# Advent of Code 2023 Day 14
#

import sys
from collections import defaultdict

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.common import log
from aoc.grid import CharGrid

Input = CharGrid
Output1 = int
Output2 = int


TEST = """\
O....#....
O.OO#....#
.....##...
OO.#O....O
.O.....O#.
O.#..O.#.#
..O..#O..O
.......O..
#....###..
#OO..#....
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    grid: CharGrid
    height: int
    width: int

    def parse_input(self, input_data: InputData) -> Input:
        return CharGrid.from_strings([line for line in input_data])

    def tilt_up(self) -> set[tuple[int, int]]:
        os = set[tuple[int, int]]()
        for c in range(self.width):
            last_cube = 0
            count = 0
            for r in range(self.height):
                val = self.grid.values[r][c]
                if val == "#":
                    last_cube = r + 1
                    count = 0
                elif val == "O":
                    new_row = last_cube + count
                    os.add((new_row, c))
                    count += 1
        return os

    def tilt_down(self) -> set[tuple[int, int]]:
        os = set[tuple[int, int]]()
        for c in range(self.width):
            last_cube = self.height - 1
            count = 0
            for r in range(self.height - 1, -1, -1):
                val = self.grid.values[r][c]
                if val == "#":
                    last_cube = r - 1
                    count = 0
                elif val == "O":
                    new_row = last_cube - count
                    os.add((new_row, c))
                    count += 1
        return os

    def tilt_left(self) -> set[tuple[int, int]]:
        os = set[tuple[int, int]]()
        for r in range(self.height):
            last_cube = 0
            count = 0
            for c in range(self.width):
                val = self.grid.values[r][c]
                if val == "#":
                    last_cube = c + 1
                    count = 0
                elif val == "O":
                    new_col = last_cube + count
                    os.add((r, new_col))
                    count += 1
        return os

    def tilt_right(self) -> set[tuple[int, int]]:
        os = set[tuple[int, int]]()
        for r in range(self.height):
            last_cube = self.width - 1
            count = 0
            for c in range(self.width - 1, -1, -1):
                val = self.grid.values[r][c]
                if val == "#":
                    last_cube = c - 1
                    count = 0
                elif val == "O":
                    new_col = last_cube - count
                    os.add((r, new_col))
                    count += 1
        return os

    def spin_cycle(self) -> set[tuple[int, int]]:
        os = self.tilt_up()
        self.redraw(os)
        os = self.tilt_left()
        self.redraw(os)
        os = self.tilt_down()
        self.redraw(os)
        os = self.tilt_right()
        self.redraw(os)
        return os

    def redraw(self, os: set[tuple[int, int]]) -> None:
        for r in range(self.height):
            for c in range(self.width):
                val = self.grid.values[r][c]
                if (r, c) in os:
                    self.grid.values[r][c] = "O"
                elif val != "#":
                    self.grid.values[r][c] = "."

    def calc_load(self, os: set[tuple[int, int]]) -> int:
        return sum(self.height - row for row, _ in os)

    def part_1(self, grid_in: Input) -> Output1:
        self.grid = grid_in
        self.height = self.grid.get_height()
        self.width = self.grid.get_width()
        os = self.tilt_up()
        self.redraw(os)
        return self.calc_load(os)

    def part_2(self, grid_in: Input) -> Output2:
        self.grid = grid_in
        self.height = self.grid.get_height()
        self.width = self.grid.get_width()
        d = defaultdict[frozenset[tuple[int, int]], list[int]](list)
        i = 0
        while True:
            if i % 100 == 0:
                log(i)
            i += 1
            os = self.spin_cycle()
            key = frozenset(os)
            d[key].append(i)
            if i > 100:
                cycle = d[key]
                if len(cycle) < 2:
                    continue
                period = cycle[1] - cycle[0]
                loops = (1_000_000_000 - i) // period
                left = 1_000_000_000 - (i + loops * period)
                log(f"{i=}, {d[key]=}, {period=}, {loops=}, {left=}")
                assert i + loops * period + left == 1_000_000_000
                for i in range(left):
                    os = self.spin_cycle()
                return self.calc_load(os)

    @aoc_samples(
        (
            ("part_1", TEST, 136),
            ("part_2", TEST, 64),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2023, 14)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2023_14.java
import static com.github.pareronia.aoc.Utils.last;
import static java.util.stream.Collectors.toSet;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2023_14
        extends SolutionBase<CharGrid, Integer, Integer> {
    
    private AoC2023_14(final boolean debug) {
        super(debug);
    }
    
    public static AoC2023_14 create() {
        return new AoC2023_14(false);
    }
    
    public static AoC2023_14 createDebug() {
        return new AoC2023_14(true);
    }
    
    @Override
    protected CharGrid parseInput(final List<String> inputs) {
        return CharGrid.from(inputs);
    }
    
    private Set<Cell> tiltUp(final CharGrid grid) {
        final Set<Cell> os = new HashSet<>();
        for (final int c : grid.colIndices()) {
            Cell cell = Cell.at(0, c);
            final Iterator<Cell> it = grid.getCellsS(cell).iterator();
            int lastCube = 0;
            int count = 0;
            while (true) {
                if (grid.getValue(cell) == '#') {
                    lastCube = cell.getRow() + 1;
                    count = 0;
                } else if (grid.getValue(cell) == 'O') {
                    os.add(Cell.at(lastCube + count, c));
                    count++;
                }
                if (!it.hasNext()) {
                    break;
                }
                cell = it.next();
            }
        }
        return os;
    }
    
    private void redraw(final CharGrid grid, final Set<Cell> os) {
        grid.getCells().forEach(cell -> {
            final char val = grid.getValue(cell);
            if (os.contains(cell)) {
                grid.setValue(cell, 'O');
            } else if (val != '#') {
                grid.setValue(cell, '.');
            }
        });
    }
    
    private int calcLoad(final CharGrid grid) {
        return grid.getAllEqualTo('O')
            .mapToInt(o -> grid.getHeight() - o.getRow())
            .sum();
    }
    
    private CharGrid spinCycle(CharGrid grid) {
        for (int i = 0; i < 4; i++) {
            this.redraw(grid, this.tiltUp(grid));
            grid = grid.rotate();
        }
        return grid;
     }

    @Override
    public Integer solvePart1(final CharGrid gridIn) {
        final CharGrid grid = gridIn.doClone();
        this.redraw(grid, this.tiltUp(grid));
        return this.calcLoad(grid);
    }
    
    @Override
    public Integer solvePart2(final CharGrid grid) {
        CharGrid g = grid.doClone();
        final Map<Set<Cell>, List<Integer>> map = new HashMap<>();
        final int total = 1_000_000_000;
        int cycles = 0;
        while (true) {
            cycles++;
            g = this.spinCycle(g);
            final Set<Cell> os = g.getAllEqualTo('O').collect(toSet());
            map.computeIfAbsent(os, k -> new ArrayList<>()).add(cycles);
            if (cycles > 100 && map.getOrDefault(os, List.of()).size() > 1) {
                final List<Integer> cycle = map.get(os);
                final int period = last(cycle, 1) - last(cycle, 2);
                final int loops = Math.floorDiv(total - cycles, period);
                final int left = total - (cycles + loops * period);
                log("cycles: %d, cycle: %s, period: %d, loops: %d, left: %d"
                        .formatted(cycles, cycle, period, loops, left));
                assert cycles + loops * period + left == total;
                for (int i = 0; i < left; i++) {
                    g = this.spinCycle(g);
                }
                return this.calcLoad(g);
            }
            if (cycles > 1000) {
                throw new IllegalStateException("Unsolvable");
            }
        }
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "136"),
        @Sample(method = "part2", input = TEST, expected = "64"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2023_14.create().run();
    }

    private static final String TEST = """
            O....#....
            O.OO#....#
            .....##...
            OO.#O....O
            .O.....O#.
            O.#..O.#.#
            ..O..#O..O
            .......O..
            #....###..
            #OO..#....
            """;
}
